doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py

```diff
@@ -173,6 +164,18 @@ def training_function(kwargs: dict):
     special_tokens = kwargs.get("special_tokens", [])
     model_id = config["model_name"]
 
+    # We need to download the model weights on this machine if they don't exit.
+    # We need to acquire a lock to ensure that only one process downloads the model
+    bucket_uri = get_mirror_link(model_id)
+    download_path = get_download_path(model_id)
+    base_path = Path(download_path).parent
+    base_path.mkdir(parents=True, exist_ok=True)
+    lock_file = str(base_path / f'{model_id.replace("/",  "--")}.lock')
+    with FileLock(lock_file):
+        download_model(
+            model_id=model_id, bucket_uri=bucket_uri, s3_sync_args=["--no-sign-request"]
+        )
```

Comment:
Do you need some sort of de-dup logic to only download the model if it doesn't exist here?

Reply:
The logic is implemented by aws cli. It is a sync operation. If it already exists it won't sync. It's smarter than if it exists, it also detects file changes.